ragstack-e2e-tests/e2e_tests/conftest.py

```diff
@@ -42,15 +53,108 @@ class AstraRef:
+            raise ValueError("Not an AstraDB test")
+        env = os.environ.get("ASTRA_DB_ENV", "prod").lower()
+        return AstraRef(
+            token=get_required_env("ASTRA_DB_TOKEN"),
+            api_endpoint=get_required_env("ASTRA_DB_ENDPOINT"),
+            collection=get_required_env("ASTRA_TABLE_NAME"),
+            id=get_required_env("ASTRA_DB_ID"),
+            env=env,
+        )
+
+    def ensure_implements_astradb(self):
+        if not self.is_astradb():
+            pytest.skip("Skipping test because Astra is not configured")
+
+    def ensure_implements_cassandra(self):
+        pass
+
+    def get_astra_client(self) -> LibAstraDB:
+        if not self.is_astradb():
+            raise ValueError("Not an AstraDB test")
+        astra_ref = self.get_astra_ref()
+        return LibAstraDB(
+            token=astra_ref.token,
+            api_endpoint=astra_ref.api_endpoint,
+        )
+
+    def get_cassandra_session(self) -> Session:
+        if not self.is_dse():
+            raise ValueError("Not running in dse mode")
+        return self.cassandra_session
+
+    def after_test(self):
+        self.test_table_name = None
+        if self.is_astradb():
+            astra_ref = self.get_astra_ref()
+            delete_astra_collection(astra_ref)
+            delete_all_astra_collections_with_client(
+                LibAstraDB(
+                    token=astra_ref.token,
+                    api_endpoint=astra_ref.api_endpoint,
+                )
+            )
+        pass
+
+    def before_test(self, implementation):
+        if self.is_astradb():
+            self.test_table_name = get_required_env("ASTRA_TABLE_NAME")
+            astra_ref = self.get_astra_ref()
+            astra_db_client = LibAstraDB(
+                token=astra_ref.token,
+                api_endpoint=astra_ref.api_endpoint,
+            )
+            delete_all_astra_collections_with_client(astra_db_client)
+
+            if implementation == "cassandra":
+                # to run cassandra implementation over astra
+                if astra_ref.env == "dev":
+                    bundle_url_template = "https://api.dev.cloud.datastax.com/v2/databases/{database_id}/secureBundleURL"
+                    cassio.init(
+                        token=astra_ref.token,
+                        database_id=astra_ref.id,
+                        bundle_url_template=bundle_url_template,
+                    )
+                else:
+                    cassio.init(token=astra_ref.token, database_id=astra_ref.id)
+        elif self.is_dse():
+            self.test_table_name = "table_" + str(random.randint(0, 1000000))
+            if self.cassandra_container is None:
+                self.cassandra_container = CassandraContainer()
+                self.cassandra_container.start()
+                logging.info("Cassandra container started")
+            else:
+                logging.info("Cassandra container already started")
```

Comment:
Is this an expected scenario? Will we create one container per test run and reuse?

Reply:
yes the startup of the container takes ~1min. We drop and recreate the keyspace for each test case